File: MST/Kruskal/P2.cpp
//MST_Kruskal -0,1
// Created by Sifat Ali on 5/4/2024.
//
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Disjoint_set {
    int *parent;
    int *rank;
    int value;

public:
    Disjoint_set(int value) {
        this->value = value;
        parent = new int[value];
        rank = new int[value];
        for (int i = 0; i < value; i++) {
            parent[i] = i;
            rank[i] = 0; // Initialize rank of each set to 0
        }
    }

    Disjoint_set() {
        delete[] parent;
        delete[] rank;
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void union_sets(int x, int y) {
        int xroot = find(x);
        int yroot = find(y);

        if (xroot == yroot) return; // Already in the same set

        if (rank[xroot] < rank[yroot]) {
            parent[xroot] = yroot;
        } else if (rank[xroot] > rank[yroot]) {
            parent[yroot] = xroot;
        } else {
            parent[yroot] = xroot;
            rank[xroot]++;
        }
    }
};

// Structure to represent an edge
struct Edge {
    int src, dest;
    int weight;
};

// Comparator function to sort edges by weight
bool compareEdges(Edge a, Edge b) {
    return a.weight < b.weight;
}

// Function to find Minimum Spanning Tree using Kruskal's algorithm
vector<Edge> MST_Kruskal(vector<Edge> edges, int V) {
    vector<Edge> result;
    Disjoint_set ds(V);
    int totalWeight = 0; // Variable to store the total weight of the MST

    // Sort edges by weight
    sort(edges.begin(), edges.end(), compareEdges);

    // Iterate through all sorted edges
    for (auto& edge : edges) {
        int src = edge.src;
        int dest = edge.dest;

        // Find sets of source and destination vertices
        int x = ds.find(src);
        int y = ds.find(dest);

        // If including this edge does not cause a cycle, include it in result and union the sets
        if (x != y) {
            result.push_back(edge);
            ds.union_sets(x, y);
            totalWeight += edge.weight; // Increment the total weight
        }
    }

    cout << "Total weight of the Minimum Spanning Tree: " << totalWeight << endl;

    return result;
}

int main() {
    // Example usage for undirected graph
    int V = 9; // Number of vertices

    vector<Edge> edges = {{0, 1, 4},
                          {0, 7, 8},
                          {1, 2, 8},
                          {1, 7, 11},
                          {2, 3, 7},
                          {2, 5, 4},
                          {2, 8, 2},
                          {3, 4, 9},
                          {3, 5, 14},
                          {4, 5, 10},
                          {5, 6, 2},
                          {6, 7, 1},
                          {6, 8, 6},
                          {7, 8, 1}};

    vector<Edge> MST = MST_Kruskal(edges, V);

    // Output the edges of MST
    cout << "Edges in the Minimum Spanning Tree (Undirected Graph):\n";
    for (auto& edge : MST) {
        cout << edge.src << " - " << edge.dest << " : " << edge.weight << endl;
    }

    return 0;
}

File: Disjoint Set/P2.py
class DisjointSet:
    def __init__(self, size):
        self.parent = [i for i in range(size)]
        self.rank = [0] * size

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union_sets(self, x, y):
        xroot = self.find(x)
        yroot = self.find(y)

        if xroot == yroot:
            return

        if self.rank[xroot] < self.rank[yroot]:
            self.parent[xroot] = yroot
        elif self.rank[yroot] < self.rank[xroot]:
            self.parent[yroot] = xroot
        else:
            self.parent[yroot] = xroot
            self.rank[xroot] += 1

    def is_cyclic(self, x, y):
        return self.find(x) == self.find(y)


# Example usage
ds = DisjointSet(5)

ds.union_sets(0, 2)
ds.union_sets(1, 4)
ds.union_sets(3, 2)

print("Representative of 0:", ds.find(0))
print("Representative of 1:", ds.find(1))
print("Representative of 2:", ds.find(2))
print("Representative of 3:", ds.find(3))

print("Is there a cycle between 3 and 4:", ds.is_cyclic(3, 4))
print("Is there a cycle between 1 and 4:", ds.is_cyclic(1, 4))
